rust/lance-index/src/vector/hnsw.rs

```diff
@@ -42,25 +167,83 @@ impl Debug for HNSW {
         write!(
             f,
             "HNSW(max_layers: {}, metric={})",
-            self.layers.len(),
+            self.levels.len(),
             self.metric_type
         )
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+struct SchemaIndexMetadata {
+    #[serde(rename = "type")]
+    type_: String,
+
+    metric_type: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct HnswMetadata {
+    entry_point: u32,
+    level_offsets: Vec<usize>,
+}
+
 impl HNSW {
-    fn from_builder(
-        layers: Vec<Arc<dyn Graph>>,
-        entry_point: u32,
-        metric_type: MetricType,
-    ) -> Self {
+    /// Load the HNSW graph from a [FileReader].
+    ///
+    /// Parameters
+    /// ----------
+    /// reader : &FileReader
+    /// vector_storage : Arc<dyn VectorStorage>
+    pub async fn load(reader: &FileReader, vector_storage: Arc<dyn VectorStorage>) -> Result<Self> {
+        let schema = reader.schema();
+        let mt = if let Some(index_metadata) = schema.metadata.get("lance:index") {
+            let index_metadata: SchemaIndexMetadata = serde_json::from_str(index_metadata)?;
+            if index_metadata.type_ != "HNSW" {
+                return Err(Error::Index {
+                    message: "index type is not HNSW".to_string(),
+                    location: location!(),
+                });
+            }
+            MetricType::try_from(index_metadata.metric_type.as_str())?
+        } else {
+            return Err(Error::Index {
+                message: "index metadata not found in the schema".to_string(),
+                location: location!(),
+            });
+        };
+        let hnsw_metadata: HnswMetadata =
+            serde_json::from_str(schema.metadata.get("lance:hnsw").ok_or_else(|| {
+                Error::Index {
+                    message: "hnsw metadata not found in the schema".to_string(),
+                    location: location!(),
+                }
+            })?)?;
+
+        let mut levels = vec![];
+        for i in 0..hnsw_metadata.level_offsets.len() - 1 {
+            let start = hnsw_metadata.level_offsets[i];
+            let end = hnsw_metadata.level_offsets[i + 1];
```

Comment:
So this is why you were thinking of a skip table for lists?

Reply:
The on disk data has this schema

```
{
   "neighbors": List<u32>,
  "pointer": u64
}
```

Assuming we can have on disk data be sorted nicely, we should have quite some read like `read_rows(10...20)` for paginations.